### PR TITLE
Dylan impl

### PR DESCRIPTION
**Pull request by webbdylan**
stack implementation working

### CODE REVIEW

File: data_structures/src/Stack/Stack.cs [hunk 1]
```diff
@@ -6,44 +6,71 @@ namespace ADT
 {
     public class Stack<T> : IStack<T>
     {
+        private T[] _stack { get; set; }
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: Call this `items` to make it clear what it is. Your internal representation of the data is not equal to the behavior of your data structure - in this case, a stack.

File: data_structures/src/Stack/Stack.cs [hunk 2]
```diff
@@ -6,44 +6,71 @@ namespace ADT
...
         public IEnumerator<T> GetEnumerator()
         {
-            throw new NotImplementedException();
+             return (IEnumerator<T>)_stack.GetEnumerator(); 
         }
 
         public bool IsEmpty()
         {
-            throw new NotImplementedException();
+            return this.Size() > 0 ? false : true;
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: No need for the `this` keyword, it is noisy and redundant in a language like c#. You also don't need this ternary expression; `return this.Size() > 0` is sufficient.

File: data_structures/src/Stack/Stack.cs [hunk 3]
```diff
@@ -6,44 +6,71 @@ namespace ADT
...
         public T Peek()
         {
-            throw new NotImplementedException();
+            if(!this.IsEmpty())
+            {
+                return _stack[_lastIndex -1];
+            }
+            else
+            {
+                throw new Exception("Stack is empty- Cannot peek.");
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: Nitpick: change `"Stack is empty- Cannot peek."` to `"Stack is empty - Cannot peek."`

File: data_structures/src/Stack/Stack.cs [hunk 4]
```diff
@@ -6,44 +6,71 @@ namespace ADT
...
+            }
         }
 
         public T Pop()
         {
-            throw new NotImplementedException();
+            if(!this.IsEmpty())
+            {
+                var returnItem =  _stack[_lastIndex -1];
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: You have a method, `Peek` that does most of this for you, why not use it??? 😉

File: data_structures/src/Stack/Stack.cs [hunk 5]
```diff
@@ -6,44 +6,71 @@ namespace ADT
...
+                _lastIndex--;
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: Don't forget to also erase the data in `_lastIndex` before decrementing it. Something like `_stack[lastIndex--] = default(T)` should do (note, I have not tried this).

File: data_structures/src/Stack/Stack.cs [hunk 6]
```diff
@@ -6,44 +6,71 @@ namespace ADT
...
+            }
         }
 
         public void Push(T item)
         {
-            throw new NotImplementedException();
+            if(_stack.Length > _lastIndex)
+            {
+                _stack[_lastIndex] = item;
+                _lastIndex++;
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: You can shorten this line by doing `_stack[_lastIndex++] = item;` Also, if you only have one statement following an if else block, no need for braces, they are noisy.

File: data_structures/src/Stack/Stack.cs [hunk 7]
```diff
@@ -6,44 +6,71 @@ namespace ADT
...
+            else
+            {
+                throw new Exception("Cannot Push- stack is full.");
+            }
         }
 
         public int Size()
         {
-            throw new NotImplementedException();
+            return _lastIndex;
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs]: It seems you are using `_lastIndex` as a way to know where the top of the stack is, why not name it accordingly?

File: data_structures/test/Stack.Tests/Tests.cs [hunk 8]
```diff
@@ -29,7 +29,7 @@ public void Push()
         public void Push_ThrowsWhenFull()
         {
             //Given
-            for (var i = 0; i < 10; i++)
```
Comment (iAmWillShepherd) [inline data_structures/test/Stack.Tests/Tests.cs:32]: Nice catch! This is what happens when you code at 🌙 !

**Review by iAmWillShepherd — CHANGES_REQUESTED** [review]

**Comment by iAmWillShepherd** [comment]
Overall, this is good stuff 👍 . Did you read the [notes](http://algs4.cs.princeton.edu/13stacks/) on the uses of stacks? If not, I'd recommend it. We will be continuing with the stack implementation to make it more useful! I promise it will get more exciting!!!

**Comment by webbdylan** [comment]
I skimmed over the article, but will do it again. I'll also get to your
corrections tomorrow sometime.

I did notice your suggestion on the 'this.'... i do agree, must've had
typescript on the brain.

Gotta run, catching up on some lost work hours at the moment...

-Dylan

**Comment by iAmWillShepherd** [comment]
Yea I figured the `this` thing was from TS, I deal with it daily. I'll be on a plane to LA tomorrow morning so I'll review it once I land. Just a heads up, the next task will be to implement a dynamically sized stack! Something to think about.

File: data_structures/src/Stack/Stack.cs (hunk 3 above)
Comment (webbdylan) [reply data_structures/src/Stack/Stack.cs]: ive been using dashes wrong my entire life... lol

File: data_structures/src/Stack/Stack.cs [hunk 9]
```diff
@@ -6,44 +6,61 @@ namespace ADT
...
+            return Size() == 0;
         }
 
         public T Peek()
         {
-            throw new NotImplementedException();
+            if(!IsEmpty())            
+                return _items[_topOfStack -1];            
+            else            
+                throw new Exception("Stack is empty - Cannot peek.");
```
Comment (iAmWillShepherd) [inline data_structures/src/Stack/Stack.cs:32]: Just wanted to point out that it's better style to check for error conditions and return as soon as possible. So swapping the conditional to:

```cs
public T Peek()
{
	if(IsEmpty())
	    throw new Exception("Stack is empty - Cannot peek.");

	return _items[_topOfStack -1];
}
```
Basically the else should only be used for things that matter.

**Review by iAmWillShepherd — APPROVED** [review]

**Pull request merged**